### PR TITLE
[ConvectionDiffusionApplication] Modernize application. Adding solver and analysis stage

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

**Comment by loumalouomega** [comment]
@rubenzorrilla and @RiccardoRossi here the thermic cylinder

[cylinder_thermic.gid.tar.gz](https://github.com/KratosMultiphysics/Kratos/files/2055530/cylinder_thermic.gid.tar.gz)

File: applications/convection_diffusion_application/custom_utilities/pure_convection_CrankN_tools.h
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/custom_utilities/pure_convection_CrankN_tools.h]: LareSe, with just 1s

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: Why do you need to echo success?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py:29]: you are passing the model, are u sure this is derived from Release 6.0 and not from the master?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: i would not flush here.

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:175]: else raise no?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:425]: what is this function? i would say it should not be here...

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:499]: nodally ->nodal

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:502]: what are u doing here? is it something new?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py:45]: mechanical?

File: applications/convection_diffusion_application/python_scripts/convergence_criteria_factory.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/convergence_criteria_factory.py:17]: use more descriptive names, D_RT and the equivalent variables suggest nothing

File: applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py
Comment (RiccardoRossi) [inline applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py:1]: what is this solver?

File: kratos/processes/replace_elements_and_condition_process.cpp
Comment (RiccardoRossi) [inline kratos/processes/replace_elements_and_condition_process.cpp]: this file should not be modified in this PR

File: kratos/includes/kratos_application.h
Comment (RiccardoRossi) [inline kratos/includes/kratos_application.h]: this file should not be modified in this PR

File: kratos/processes/replace_elements_and_condition_process.h
Comment (RiccardoRossi) [inline kratos/processes/replace_elements_and_condition_process.h]: this file should not be modified in this Pr

File: kratos/python_scripts/read_materials_process.py
Comment (RiccardoRossi) [inline kratos/python_scripts/read_materials_process.py:169]: do you need to modify this file? i think there is already a PR with this proposed modification. We'll consider this change in that PR

File: kratos/sources/variables.cpp
Comment (RiccardoRossi) [inline kratos/sources/variables.cpp]: i believe that this file should not be modified as a part of this PR

File: kratos/tests/processes/test_replace_elements_and_condition_process.cpp
Comment (RiccardoRossi) [inline kratos/tests/processes/test_replace_elements_and_condition_process.cpp:1]: this file should not be modified (or added) in this PR, since the replace function should also not be changed

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]

File: applications/convection_diffusion_application/custom_utilities/pure_convection_CrankN_tools.h
Comment (loumalouomega) [reply applications/convection_diffusion_application/custom_utilities/pure_convection_CrankN_tools.h]: OK

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: It is the same as the structural analysis

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py:29]: I am using the Release 6.0 as reference, exactly the same as the structural

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: Exactly the same as the structural

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:175]: what?

File: kratos/includes/kratos_application.h
Comment (loumalouomega) [reply kratos/includes/kratos_application.h]: OK

File: kratos/processes/replace_elements_and_condition_process.h
Comment (loumalouomega) [reply kratos/processes/replace_elements_and_condition_process.h]: There is a PR specific for this for the master

https://github.com/KratosMultiphysics/Kratos/pull/2221

Comment (loumalouomega) [reply kratos/processes/replace_elements_and_condition_process.h]: I added all together for the Release PR

File: kratos/python_scripts/read_materials_process.py
Comment (loumalouomega) [reply kratos/python_scripts/read_materials_process.py:169]: This is already merged in the master

https://github.com/KratosMultiphysics/Kratos/pull/2217

File: kratos/tests/processes/test_replace_elements_and_condition_process.cpp
Comment (loumalouomega) [reply kratos/tests/processes/test_replace_elements_and_condition_process.cpp:1]: There is a PR specific for this for the master

#2221

File: kratos/sources/variables.cpp
Comment (loumalouomega) [reply kratos/sources/variables.cpp]: Ok

**Comment by loumalouomega** [comment]
Done @RiccardoRossi, as I said for the master #2221 is waiting for approval and #2217
 is already merged. The other comments will imply to modify code in the StructuralMechanics too

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (philbucher) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: I thought it would be good to know, but feel free to remove it

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (philbucher) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: Will be removed in #2135

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (philbucher) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:425]: Parameters in derived classes are not validated in the baseclass

Have a look at how this is being used in Structural

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:175]: If the var name is empty you should raise an exception no?

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:425]: I ll pass on this for now, but we should find a better solution. And definitely we should not replicate the code in every solver

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: Then let a remove it

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_stationary_solver.py:45]: This is not a mechanical scheme

File: kratos/tests/processes/test_replace_elements_and_condition_process.cpp
Comment (RiccardoRossi) [reply kratos/tests/processes/test_replace_elements_and_condition_process.cpp:1]: I don don t see any technical reason for backporting this to release. And it is not approved in master so it should not go in

**Comment by RiccardoRossi** [comment]
Modifications can be applied to convection diffusion without having to be applied onto structural mechanics as well...

File: kratos/tests/processes/test_replace_elements_and_condition_process.cpp
Comment (loumalouomega) [reply kratos/tests/processes/test_replace_elements_and_condition_process.cpp:1]: It is exactly the same, but with a test added

**Comment by loumalouomega** [comment]
OK

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: OK

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_analysis.py]: OK

File: kratos/processes/replace_elements_and_condition_process.h
Comment (RiccardoRossi) [reply kratos/processes/replace_elements_and_condition_process.h]: But it is not needed here, so do not backport.

File: applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py
Comment (loumalouomega) [reply applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py:175]: Nope, it is already overrided in the init, but in case you are interested in not set that variable you can use "" instead

File: applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py
Comment (RiccardoRossi) [reply applications/convection_diffusion_application/python_scripts/proposal_convection_diffusion_solver.py:1]: Ping

**Comment by loumalouomega** [comment]
@RiccardoRossi Done, reverted ReplaceElementsAndConditionsProcess. I keep the test, tests are good

**Comment by loumalouomega** [comment]
Besides, #2221 merged

**Comment by loumalouomega** [comment]
If we are OK, I will PR this to the master

**Comment by roigcarlo** [comment]
I think this is the last PR pending to 6.0.0.
When this is merged please ping me so I can start building the binaries

**Comment by loumalouomega** [comment]
Anyone can approve @RiccardoRossi ?

File: kratos/python_scripts/read_materials_process.py
Comment (RiccardoRossi) [reply kratos/python_scripts/read_materials_process.py:169]: ok i am accepting this

File: kratos/tests/processes/test_replace_elements_and_condition_process.cpp
Comment (RiccardoRossi) [reply kratos/tests/processes/test_replace_elements_and_condition_process.cpp:1]: I am accepting this since we are late for the release. next time i will not allow backporting a feature that is not needed.

**Review by RiccardoRossi — APPROVED** [review]
i am accepting even if some of my comments were addressed. I will block PR to master to ensure that all the others are correctly addressed

**Comment by RiccardoRossi** [comment]
@roigcarlo i approved to make possible to build the release

**Pull request merged**
